clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesV2CachingAgentDispatcher.java

```diff
@@ -44,17 +48,52 @@
   @Autowired
   private KubectlJobExecutor jobExecutor;
 
+  @Autowired
+  private KubernetesResourcePropertyRegistry propertyRegistry;
+
   @Override
   public List<KubernetesCachingAgent> buildAllCachingAgents(KubernetesNamedAccountCredentials credentials) {
     return IntStream.range(0, credentials.getCacheThreads())
         .boxed()
-        .map(i -> new ArrayList<KubernetesCachingAgent>(Arrays.asList(
-            new KubernetesNetworkPolicyCachingAgent(credentials, jobExecutor, objectMapper, registry, i, credentials.getCacheThreads()),
-            new KubernetesPodCachingAgent(credentials, jobExecutor, objectMapper, registry, i, credentials.getCacheThreads()),
-            new KubernetesReplicaSetCachingAgent(credentials, jobExecutor, objectMapper, registry, i, credentials.getCacheThreads()),
-            new KubernetesServiceCachingAgent(credentials, jobExecutor, objectMapper, registry, i, credentials.getCacheThreads())
-        )))
-        .flatMap(Collection::stream)
+        .map(i -> propertyRegistry.values()
+            .stream()
+            .map(KubernetesResourceProperties::getHandler)
+            .map(KubernetesHandler::cachingAgentClass)
+            .filter(Objects::nonNull)
+            .map(c -> {
+                  try {
+                    return c.getDeclaredConstructor(
+                        KubernetesNamedAccountCredentials.class,
+                        KubectlJobExecutor.class,
+                        ObjectMapper.class,
+                        Registry.class,
+                        int.class,
+                        int.class
+                    );
+                  } catch (NoSuchMethodException e) {
+                    log.warn("Missing canonical constructor", e);
+                    return null;
+                  }
+                }
+            )
+            .map(c -> {
+                  try {
+                    return (KubernetesV2CachingAgent) c.newInstance(
```

Comment:
Won't you hit an NPE on `c.newInstance` if you really returned `null` in the last `.map`? ...or do I not know Java?

Reply:
No I think you know more than I do - will fix in the morning